### PR TITLE
[RF-DOCS] Action Mailer Guide [ci-skip]

### PR DESCRIPTION
**Pull request by bhumi1102**
### Motivation / Background

This Pull Request is for updating and improving the "Action Mailer Basics" Rails Guide documentation.

### Testing

- [x] Run `guides:generate`
- [x] Run `guides:lint` 
- [x] Run rubocop on code samples

### Details

Here are some of the things addressed from this Guide.

- [X] Mailers now actually inherit from a generated ApplicationMailer, which in turn inherit from ActionMailer::Base.
- [X] The output to generating a mailer might be different depending on whether the base ApplicationMailer and layout have already been generated, which will already happen when creating a new app, or if they're not there the first time the mailer is generated.
- [X] Maybe the generator should be shown by passing an action name directly? (or as a second example or subsection?) bin/rails generate mailer User welcome_email
- [X] Complete List of Action Mailer should be linked. (in fact, maybe we don't even "need" this to be a section, but rather part of the text when introducing the other mailer methods? And then a full section on Attachments)
- [X] The HTML view has a complete layout with DOCTYPE/head/body, but it should be using the application mailer's layout, so that example will likely double the HTML, it is incorrect.
- [X] The guide shows how to build a multipart email (HTML + Text), but doesn't really explain why, what are the benefits, if it's really needed, etc., might be a worth addition to consider. There's a section later in the guide that's about "multipart emails", that we could link to from the top, and perhaps expand a bit on that.
- [X] I wonder if we should instead show an example of sending the email via console, after we have the mailer and associated model? And then proceed with showing the controller version.
- [X] Link to Active Job guide, there's a mention when explaining how to call the mailer, and a Note. That note should likely be simplified and linked instead. (less duplicated content, no need to mention Sidekiq,etc. elsewhere, etc.)
- [X]  If we do show how to use via console, it'd be a good introduction to start with deliver_now, and then explain deliver_later + Active Job afterwards.
- [X] Auto Encoding Header Values seems like a topic/section that could be pushed down, it's not a common need.
Decided to remove as it is an extraneous detail.
- [X] Attachments could be a larger subsection, right now it's a 3rd level section that doesn't show in the sections list on the right.
- [X] "Action Mailer 3.0" no need to mention older versions any longer.
- [X] Under the Layouts section, it doesn't mention where the layout should be located. (app/views/layouts), and it should probably mention the default mailer layout now too. (see other comment about about the email template containing the full HTML tags)
- [X] There's a section about "Dynamic Delivery Options" that shows how to pass SMTP & etc. configs, but the configuration section is later in the guide, this could be reorganized to better flow, or at least linked.
- [X] There's some information about views, layouts, rendering without a template, etc., scattered, maybe it makes sense to group these related pieces a little bit better.
- [X] Under the configuration section, deliveries is not really a configuration but something to use in dev/test., not sure it's where it should be mentioned.
- [X] A new section about rescuing errors could be added too. [Here's the API](https://github.com/rails/rails/blob/3efae445196da943c0532a00f6dfbafc74ae879d/actionmailer/lib/action_mailer/base.rb#L353-L374) for reference.
- [X] No mention of I18n, maybe could have at very brief section or paragraph/note somewhere that points to the [I18n guide on the matter](https://edgeguides.rubyonrails.org/i18n.html#translations-for-action-mailer-e-mail-subjects).

### CODE REVIEW

File: guides/source/action_mailer_basics.md [hunk 1]
```diff
@@ -3,155 +3,143 @@
...
+* How to preview emails and test your Action Mailer classes.
 
 --------------------------------------------------------------------------------
 
 What is Action Mailer?
 ----------------------
 
-Action Mailer allows you to send emails from your application using mailer classes
-and views.
+Action Mailer allows you to send emails from your Rails application. It's one of the two email related components in the Rails framework. Action Mailer, which is for sending emails and [Action Mailbox](action_mailbox_basics.html), which deals with receiving emails.
```
Comment (kimadactyl) [inline guides/source/action_mailer_basics.md:21]:
```suggestion
Action Mailer allows you to send emails from your Rails application. It's one of two email-related components in the Rails framework: 

1. Action Mailer, for sending emails
2. [Action Mailbox](action_mailbox_basics.html), for receiving emails
```

Comment (kimadactyl) [reply guides/source/action_mailer_basics.md:21]: Small suggestion to make the two components ultra clear

File: guides/source/action_mailer_basics.md [hunk 2]
```diff
@@ -3,155 +3,143 @@
...
 After reading this guide, you will know:
 
-* How to send email within a Rails application.
-* How to generate and edit an Action Mailer class and mailer view.
+* How to generate and edit Action Mailer classes and mailer views.
+* How to send attachments and multipart emails.
+* How to use Action Mailer callbacks.
 * How to configure Action Mailer for your environment.
-* How to test your Action Mailer classes.
+* How to preview emails and test your Action Mailer classes.
```
Comment (bhumi1102) [inline guides/source/action_mailer_basics.md:15]: Todo: add something about CustomDeliveryJob https://x.com/rails/status/1834324357088481381

**Comment by rodrigotoledo** [comment]
It's more straightforward, and with these new points, it's easier to understand.

File: guides/source/action_mailer_basics.md (hunk 1 above)
Comment (bhumi1102) [reply guides/source/action_mailer_basics.md:21]: updated

File: guides/source/action_mailer_basics.md [hunk 3]
```diff
@@ -3,155 +3,174 @@
...
+  mailer. In this case, we use it to set the `:from` header value for all
   messages in this class. This can be overridden on a per-email basis.
 * The [`mail`][] method creates the actual email message. We use it to specify
   the values of headers like `:to` and `:subject` per email.
 
-[`default`]: https://api.rubyonrails.org/classes/ActionMailer/Base.html#method-c-default
-[`mail`]: https://api.rubyonrails.org/classes/ActionMailer/Base.html#method-i-mail
+There is also the [`headers`][] method (not used above), which is used to
+specify email headers with a hash or by calling `headers[:field_name] =
+'value'`.
```
[the remark this answers is not in the document]
Comment (bhumi1102) [reply guides/source/action_mailer_basics.md:124]: I made a note to look at the generator code to and see if I remove linebreaks from single backticks codeblocks.

File: guides/source/action_mailer_basics.md [hunk 4]
```diff
@@ -241,176 +299,170 @@ class SendWeeklySummary
...
-
-#### Adding Attachments
-
-Action Mailer makes it very easy to add attachments.
-
-* Pass the file name and content and Action Mailer and the
-  [Mail gem](https://github.com/mikel/mail) will automatically guess the
-  `mime_type`, set the `encoding`, and create the attachment.
+```
+UserMailer.with(user: user).weekly_summary
```
[the remark this answers is not in the document]
Comment (bhumi1102) [reply guides/source/action_mailer_basics.md:313]: Updated. Linter was complaining about the object syntax before but seems okay now.

File: guides/source/action_mailer_basics.md [hunk 5]
```diff
@@ -241,176 +299,170 @@ class SendWeeklySummary
...
+NOTE: The order of the parts getting inserted is determined by the
+`:parts_order` inside of the `ActionMailer::Base.default` method.
 
-The same technique works to specify a sender name:
+Multipart is also used when you send attachments with email. The `multipart`
+MIME type represents a document that's comprised of multiple component parts,
+each of which may have its own individual MIME type, such as the `text/html` and
+`text/text` parts above. The use of `multipart` type is to encapsulate multiple
+files being sent together in one transaction, such as when attaching multiple
+files to an email.
```
[the remark this answers is not in the document]
Comment (bhumi1102) [reply guides/source/action_mailer_basics.md:445]: Moved and updated and reworded this a bit.

**Comment by chaadow** [comment]
Hi @p8 is there a schedule for releasing documentation updates on edgeguides website? Thanks!

**Comment by p8** [comment]
@chaadow edgeguides are supposed to be updated after merging to main. It seems edgeguides isn't running main, I'll see if we can fix this.

**Comment by p8** [comment]
@chaadow It seems dropping support for Ruby 3.1 broke it.
The following PR hopefully makes sure edgeguides always show main:
https://github.com/rails/rails-docs-server/pull/29

**Comment by chaadow** [comment]
@p8 Thanks!

**Pull request merged**